### PR TITLE
FALCON-2191 Extension Job store changes and rest api changes for submit and submitAndScheule extension services.

### PR DESCRIPTION
**Pull request by sandeepSamudrala**

### CODE REVIEW

File: common/src/main/java/org/apache/falcon/persistence/ExtensionJobsBean.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    @Basic
+    @NotNull
+    @Column(name = "extension_name")
+    private String extensionName;
+
+    @Basic
+    @NotNull
+    @Column(name = "extension_type")
+    @Enumerated(EnumType.STRING)
+    private ExtensionType extensionType;
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/persistence/ExtensionJobsBean.java]: Why does extensionType need to be in this table?

File: common/src/main/java/org/apache/falcon/persistence/ExtensionJobsBean.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    @Basic
+    @NotNull
+    @Column(name = "creation_time")
+    private Date creationTime;
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/persistence/ExtensionJobsBean.java]: On the same lines, can you add modificationTime too? Will be useful info to give when user updates the extension job.

File: common/src/main/java/org/apache/falcon/persistence/ExtensionJobsBean.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+    @Basic
+    @NotNull
+    @Column(name = "entities")
+    private String[] entities;
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/persistence/ExtensionJobsBean.java]: This should ideally be a Pair<String, EntityType>

File: extensions/src/test/java/org/apache/falcon/extensions/jdbc/ExtensionMetaStoreTest.java [hunk 4]
```diff
@@ -57,28 +58,41 @@ public void init() {
...
         Assert.assertEquals(stateStore.getAllExtensions().size(), 1);
         //check data
-        ExtensionMetadataBean bean = stateStore.getDetail("test1");
+        ExtensionBean bean = stateStore.getDetail("test1");
         Assert.assertEquals(bean.getLocation(), "test_location");
         //delete
         stateStore.deleteExtensionsOfType(ExtensionType.TRUSTED);
         Assert.assertEquals(stateStore.getAllExtensions().size(), 0);
+
+        stateStore.storeExtensionBean("test2", "test_location", ExtensionType.CUSTOM, "test2_description");
```
Comment (pallavi-rao) [inline extensions/src/test/java/org/apache/falcon/extensions/jdbc/ExtensionMetaStoreTest.java]: Nit : Better to have it in a separate test method testExtension and testExtensionJob

File: prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java [hunk 5]
```diff
@@ -292,39 +296,76 @@ public APIResult delete(@PathParam("job-name") String jobName,
...
+        ExtensionBean extensionDetails = metaStore.getDetail(extensionName);
+        return extensionDetails.getExtensionType();
+    }
+
     @POST
     @Path("submitAndSchedule/{extension-name}")
-    @Consumes({MediaType.TEXT_XML, MediaType.TEXT_PLAIN})
+    @Consumes({MediaType.TEXT_XML, MediaType.TEXT_PLAIN, MediaType.MULTIPART_FORM_DATA})
     @Produces({MediaType.TEXT_XML, MediaType.TEXT_PLAIN, MediaType.APPLICATION_JSON})
     public APIResult submitAndSchedule(
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java]: Possible to have a common submit method that both submit and submitAndSchedule can use?

File: prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java [hunk 6]
```diff
@@ -292,39 +297,53 @@ public APIResult delete(@PathParam("job-name") String jobName,
...
         } catch (FalconException | IOException e) {
             LOG.error("Error when submitting extension job: ", e);
             throw FalconWebException.newAPIException(e, Response.Status.INTERNAL_SERVER_ERROR);
         }
         return new APIResult(APIResult.Status.SUCCEEDED, "Extension job submitted successfully");
     }
 
+    private void validateEntities(List<Entity> entities) throws FalconException {
+        for (Entity entity : entities) {
+            super.validate(entity);
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java]: Can you also ensure the entities are only of type process or feed

Comment (sandeepSamudrala) [inline prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java]: Made the changes.

**Pull request closed without merging**